pkg/test/stage.go

```diff
@@ -0,0 +1,241 @@
+// GetTimeout gets the timeout defined for the test stage.
+func (s *Stage) GetTimeout() int {
+	timeout := 10
+	if s.Assert != nil && s.Assert.Timeout != 0 {
+		timeout = s.Assert.Timeout
+	}
+	return timeout
+}
+
+// CheckResource checks if the expected resource's state in Kubernetes is correct.
+func (s *Stage) CheckResource(expected runtime.Object, namespace string) []error {
+	testErrors := []error{}
+
+	name, namespace, err := testutils.Namespaced(s.DiscoveryClient, expected, namespace)
+	if err != nil {
+		return append(testErrors, err)
+	}
+
+	gvk := expected.GetObjectKind().GroupVersionKind()
+
+	actual := &unstructured.Unstructured{}
+	actual.SetGroupVersionKind(gvk)
+
+	err = s.Client.Get(context.TODO(), client.ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}, actual)
+	if err != nil {
+		return append(testErrors, err)
+	}
+
+	expectedObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(expected)
+	if err != nil {
+		return append(testErrors, err)
+	}
+
+	if err := testutils.IsSubset(expectedObj, actual.UnstructuredContent()); err != nil {
+		diff, diffErr := testutils.PrettyDiff(expected, actual)
+		if diffErr == nil {
+			testErrors = append(testErrors, errors.New(diff))
+		} else {
+			testErrors = append(testErrors, diffErr)
+		}
+
+		testErrors = append(testErrors, fmt.Errorf("resource %s: %s", testutils.ResourceID(expected), err))
+	}
+
+	return testErrors
+}
+
+// Check checks if the resources defined in Asserts are in the correct state.
+func (s *Stage) Check(namespace string) []error {
+	testErrors := []error{}
+
+	for _, expected := range s.Asserts {
+		testErrors = append(testErrors, s.CheckResource(expected, namespace)...)
+	}
+
+	return testErrors
+}
+
+// Run runs a KUDO test stage:
+// 1. Apply all desired objects to Kubernetes.
+// 2. Wait for all of the states defined in the test stage's asserts to be true.'
+func (s *Stage) Run(namespace string) []error {
+	s.Logger.Log("starting test stage", s.String())
+
+	testErrors := s.Create(namespace)
+
+	if len(testErrors) != 0 {
+		return testErrors
+	}
+
+	for i := 0; i < s.GetTimeout(); i++ {
+		testErrors = s.Check(namespace)
+
+		if len(testErrors) == 0 {
+			break
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	if len(testErrors) == 0 {
+		s.Logger.Log("test stage completed", s.String())
+	} else {
+		s.Logger.Log("test stage failed", s.String())
+	}
+
+	return testErrors
+}
+
+// String implements the string interface, returning the name of the test stage.
+func (s *Stage) String() string {
+	return fmt.Sprintf("%d-%s", s.Index, s.Name)
+}
+
+// LoadYAML loads the resources from a YAML file for a test stage:
+// * If the YAML file is called "assert", then it contains objects to
+//   add to the test stage's list of assertions.
+// * If the YAML file is called "errors", then it contains objects that,
+//   if seen, mark a test immediately failed.
+// * All other YAML files are considered resources to create.
+func (s *Stage) LoadYAML(file string) error {
```

Comment:
(utils again?)

Reply:
It calls out to a load YAML method in utils, but is specific handling for the stages, so I think it makes sense here.